proposals/gc/MVP.md

```diff
@@ -97,20 +118,167 @@ New abbreviations are introduced for reference types in binary and text format,
+
+Validity of a module is checked under a context storing the definitions for each type. In the case of recursive types, this definition is given by a respective projection from the full type:
+```
+ctxtype ::= <deftype> | (rec <deftype>*).<i>
+```
+
+#### Auxiliary Definitions
+
 * Unpacking a storage type yields `i32` for packed types, otherwise the type itself
   - `unpacked(t) = t`
   - `unpacked(pt) = i32`
 
-TODO: Need to be able to use `i31` as a type definition.
+* Context lookup `C($t)` is extended to recursive type indices by ignoring the `rec` marker.
+  - `C(rec $t) = C($t)`
+
+* Unrolling a possibly recursive context type projects the respective item
+  - `unroll($t)                 = unroll(<ctxtype)`  iff `C($t) = <ctxtype>`
+  - `unroll(<deftype>)          = <deftype>`
+  - `unroll((rec <deftype>*).i) = (<deftype>*)[i]`
+
+* Expanding a type definition unrolls it and returns its plain definition
+  - `expand($t)                 = expand(<ctxtype)`  iff `C($t) = <ctxtype>`
+  - `expand(<ctxtype>) = <strtype>`
+    - where `unroll(<ctxttype>) = sub x* <strtype>`
+
+
+
+#### Type Validity
+
+Some of the rules define a type as `ok` for a certain index, written `ok(x)`. This controls uses of type indices as supertypes inside a recursive group: the subtype hierarchy must not be cyclic, and hence any type index used for a supertype is required to be smaller than the index `x` of the current type.
+
+* a sequence of type definitions is valid if each item is valid within the context containing the prior items
+  - `<typedef0> <typedef>* ok`
+    - iff `<typedef0> ok` and extends the context accordingly
+    - and `<typedef>* ok` under the extended context
+
+* a plain type definition is valid if its `deftype` is at its type index
+  - `<deftype> ok` and extends the context with `<deftype>`
+    - iff `<deftype> ok($t)` where `$t` is the next unused (i.e., current) type index
+
+* a recursive type definition is valid if its types are valid under the context containing all of them
+  - `rec <deftype>* ok` and extends the context with `<ctxtype>*`
+    - iff `<deftype>* ok($t)` under the extended context(!)
+    - where `x` is the next unused (i.e., current) type index
+    - and `N = |<deftype>*|-1`
+    - and `<deftype'>* = <deftype>*[$t:=rec $t, ..., $t+N:=rec $t+N]`
+    - and `<ctxtype>*  = (rec <deftype'>*).0, ..., (rec <deftype'>*).N`
+  - Note: `<deftype'>*` marks all recursive occurrences of type indices from within this group with `rec`; this is expressed here by a validation-time substitution, but an implementation could insert the annotations on the fly during decoding.
+  - Because rec type indices `rec $t`cannot occur in the source, the following are invariants that are established by these rules:
+    - (1) rec indices only occur in the context and inside a `(rec <deftype>*)`,
+    - (2) they only refer to indices from the same recursive group,
+    - (3) all internal indices within a recursive group are marked with rec,
```

Comment:
Does this imply that with rec <idx>, <idx> is in the range 0..N where there are N definitions in the rec <deftype> group?